examples/market_maker/on_chain_market_maker.vy

```diff
@@ -54,4 +54,5 @@ def tokensToEth(sell_quantity: uint256(currency_value)):
 def ownerWithdraw():
     assert self.owner == msg.sender
     self.token_address.transfer(self.owner, as_unitless_number(self.totalTokenQty))
-    selfdestruct(self.owner)
+    send(self.owner, self.balance)  # FIXME: why doesn't selfdestruct work?
```

Comment:
i think we should report this to py-evm

Reply:
Mmh, nothing that I can tell from the top of my head. The one thing that I find super valuable when debugging these things is to activate full logging output. I *think* throwing this into the script would do it.

```python
import logging
import sys
logging.basicConfig(stream=sys.stdout, level=0)
```

Then, you should see detailed tracing logs from py-evm that lets one follow all the individual opcodes that are being applied (or not applied).

I can also help to debug this but before I dive in, do you have quick instructions on how to run this? Do I just clone vyper, install dependencies and then run the tests in `tests/examples/safe_remote_purchase/test_safe_remote_purchase.py `?

If this might be related to Istanbul changes then I assume @carver might have better insight here (since he has implemented the Istanbul EIPs)

Reply:
I don't think it's related to Istanbul, because we still get the failure 2 releases before Istanbul was made the default option. Also, @iamdefinitelyahuman noticed that `SELFDESTRUCT` had no test cases in `py-evm`, so I am definitely suspect of that.

But yeah, pull this branch, `pip install -e .[test]` and then `pytest tests/examples/safe_remote_purchase/test_safe_remote_purchase.py`

Reply:
Confirmed, doesn't seem related to Istanbul.

Here's what I did so far:

I ran `pytest tests/examples/safe_remote_purchase/test_safe_remote_purchase.py -vv -k 'test_abort' -n0 --log-cli-level 0` on that branch with the `eth-tester` upgrade reverted and uncommented one of the `selfdestruct` calls. This worked (as expected). And I've put the logs here:

https://gist.github.com/cburgdorf/45ec8bffaf4ac8d03e495252fe30ffc8

Then, I've run the same with the upgraded `eth-tester`. But I also hacked the `eth-tester` code to build on a Constantinople chain just to ensure both outputs will be comparable. It still fails.

I've put the logs here:

https://gist.github.com/cburgdorf/66c94681ce54f4c04e1f53b27ec6551e

Now, I tried to see where bot logs start drift apart.

It looks like this is the last line that are the same on both sides:

`GAS CONSUMPTION: 2234828 - 10 -> 2234818 (JUMPI)`

After that line, the version that errors goes on with (Notice the `Computation Error`):

```
Level 8  eth.vm.computation.Computation:logging.py:36 OPCODE: 0x60 (PUSH1) | pc: 194
Level 8  eth.gas.GasMeter:logging.py:36 GAS CONSUMPTION: 2234818 - 3 -> 2234815 (PUSH1)
Level 8  eth.vm.computation.Computation:logging.py:36 OPCODE: 0x80 (DUP1) | pc: 196
Level 8  eth.gas.GasMeter:logging.py:36 GAS CONSUMPTION: 2234815 - 3 -> 2234812 (DUP1)
Level 8  eth.vm.computation.Computation:logging.py:36 OPCODE: 0xfd (REVERT) | pc: 197
Level 8  eth.vm.computation.Computation:logging.py:36 MEMORY: size (192 -> 0) | cost (18 -> 0)
Level 8  eth.vm.computation.Computation:logging.py:36 COMPUTATION ERROR: gas: 2235181 | from: 0x7e5f4552091a69125d5dfcb7b8c2659029395bdf | to: 0xf2e246bb76df876cef8b38ae84130f4f55de395b | value: 0 | depth: 0 | static: n | error: b''
Level 8  eth.db.storage.AccountStorageDB:logging.py:36 discard checkpoint 120
Level 8  eth.vm.state.ConstantinopleState:logging.py:36 TRANSACTION REFUND: 2234812 -> 0x7e5f4552091a69125d5dfcb7b8c2659029395bdf
Level 8  eth.vm.state.ConstantinopleState:logging.py:36 TRANSACTION FEE: 21641 -> 0x0000000000000000000000000000000000000000
Level 8  eth.db.storage.AccountStorageDB:logging.py:36 discard checkpoint 119
Level 8  eth.vm.state.ConstantinopleState:logging.py:36 TRANSACTION: sender: 0x7e5f4552091a69125d5dfcb7b8c2659029395bdf | to: 0xf2e246bb76df876cef8b38ae84130f4f55de395b | value: 0 | gas: 3374044 | gas-price: 1 | s: 1 | r: 1 | v: 37 | data-hash: 0x2fe909773c01f5ff57bee87e86da3ee355e449cd71e314ee1e7f6c6729b93717
Level 8  eth.vm.computation.Computation:logging.py:36 COMPUTATION STARTING: gas: 3352772 | from: 0x7e5f4552091a69125d5dfcb7b8c2659029395bdf | to: 0xf2e246bb76df876cef8b38ae84130f4f55de395b | value: 0 | depth 0 | static: n
```

Whereas the version that suceeds goes on with:

```
DEBUG2   eth.vm.computation.Computation:logging.py:10 OPCODE: 0x5b (JUMPDEST) | pc: 198
DEBUG2   eth.gas.GasMeter:logging.py:10 GAS CONSUMPTION: 2234818 - 1 -> 2234817 (JUMPDEST)
DEBUG2   eth.vm.computation.Computation:logging.py:10 OPCODE: 0x60 (PUSH1) | pc: 199
DEBUG2   eth.gas.GasMeter:logging.py:10 GAS CONSUMPTION: 2234817 - 3 -> 2234814 (PUSH1)
DEBUG2   eth.vm.computation.Computation:logging.py:10 OPCODE: 0x54 (SLOAD) | pc: 201
DEBUG2   eth.gas.GasMeter:logging.py:10 GAS CONSUMPTION: 2234814 - 200 -> 2234614 (SLOAD)
DEBUG2   eth.vm.computation.Computation:logging.py:10 OPCODE: 0x33 (CALLER) | pc: 202
DEBUG2   eth.gas.GasMeter:logging.py:10 GAS CONSUMPTION: 2234614 - 2 -> 2234612 (CALLER)
DEBUG2   eth.vm.computation.Computation:logging.py:10 OPCODE: 0x14 (EQ) | pc: 203
DEBUG2   eth.gas.GasMeter:logging.py:10 GAS CONSUMPTION: 2234612 - 3 -> 2234609 (EQ)
DEBUG2   eth.vm.computation.Computation:logging.py:10 OPCODE: 0x61 (PUSH2) | pc: 204
DEBUG2   eth.gas.GasMeter:logging.py:10 GAS CONSUMPTION: 2234609 - 3 -> 2234606 (PUSH2)
DEBUG2   eth.vm.computation.Computation:logging.py:10 OPCODE: 0x57 (JUMPI) | pc: 207
DEBUG2   eth.gas.GasMeter:logging.py:10 GAS CONSUMPTION: 2234606 - 10 -> 2234596 (JUMPI)
DEBUG2   eth.vm.computation.Computation:logging.py:10 OPCODE: 0x5b (JUMPDEST) | pc: 212
DEBUG2   eth.gas.GasMeter:logging.py:10 GAS CONSUMPTION: 2234596 - 1 -> 2234595 (JUMPDEST)
DEBUG2   eth.vm.computation.Computation:logging.py:10 OPCODE: 0x60 (PUSH1) | pc: 213
DEBUG2   eth.gas.GasMeter:logging.py:10 GAS CONSUMPTION: 2234595 - 3 -> 2234592 (PUSH1)
DEBUG2   eth.vm.computation.Computation:logging.py:10 OPCODE: 0x54 (SLOAD) | pc: 215
DEBUG2   eth.gas.GasMeter:logging.py:10 GAS CONSUMPTION: 2234592 - 200 -> 2234392 (SLOAD)
DEBUG2   eth.vm.computation.Computation:logging.py:10 OPCODE: 0xff (SELFDESTRUCT) | pc: 216
DEBUG2   eth.gas.GasMeter:logging.py:10 GAS CONSUMPTION: 2234392 - 5000 -> 2229392 (SELFDESTRUCT)
DEBUG2   eth.vm.computation.Computation:logging.py:10 SELFDESTRUCT: 0xf2e246bb76df876cef8b38ae84130f4f55de395b (2000000000000000000) -> 0x7e5f4552091a69125d5dfcb7b8c2659029395bdf
DEBUG2   eth.vm.computation.Computation:logging.py:10 COMPUTATION SUCCESS: from: 0x7e5f4552091a69125d5dfcb7b8c2659029395bdf | to: 0xf2e246bb76df876cef8b38ae84130f4f55de395b | value: 0 | depth: 0 | static: n | gas-used: 5789 | gas-remaining: 2229392
DEBUG2   eth.gas.GasMeter:logging.py:10 GAS REFUND: 0 + 24000 -> 24000
DEBUG2   eth.vm.state.ConstantinopleState:logging.py:10 TRANSACTION REFUND: 2242922 -> 0x7e5f4552091a69125d5dfcb7b8c2659029395bdf
DEBUG2   eth.vm.state.ConstantinopleState:logging.py:10 TRANSACTION FEE: 13531 -> 0x0000000000000000000000000000000000000000
DEBUG2   eth.vm.state.ConstantinopleState:logging.py:10 DELETING ACCOUNT: 0xf2e246bb76df876cef8b38ae84130f4f55de395b
DEBUG2   eth.vm.state.ConstantinopleState:logging.py:10 TRANSACTION: sender: 0x7e5f4552091a69125d5dfcb7b8c2659029395bdf | to: 0xf2e246bb76df876cef8b38ae84130f4f55de395b | value: 0 | gas: 1138862 | gas-price: 1 | s: 1 | r: 1 | v: 37 | data-hash: 0x2fe909773c01f5ff57bee87e86da3ee355e449cd71e314ee1e7f6c6729b93717
DEBUG2   eth.vm.computation.Computation:logging.py:10 COMPUTATION STARTING: gas: 1117590 | from: 0x7e5f4552091a69125d5dfcb7b8c2659029395bdf | to: 0xf2e246bb76df876cef8b38ae84130f4f55de395b | value: 0 | depth 0 | static: n
```

Eventually the test fails with:

>eth_utils.exceptions.ValidationError: Transaction exceeds gas limit: using 4495606, bringing total to 4495606, but limit is 4491635

Not sure if this is helpful. I'm not an expert for this low level nitty gritty stuff but I'm putting this out in hope that someone else could make more sense of it. (Summon @veox)

I guess the next thing that I'd try is to bisect the issue against different py-evm commits to find the commit that breaks the test.

Reply:
Awesome, thank you!

Reply:
Thanks @cburgdorf, that's very useful.

The fact that `vyper_flies` does `JUMPI`-`JUMPDEST` and eventually `SELFDESTRUCT`, whereas `vyper_error_constantinople` does `JUMPI`-`PUSH1`-`DUP1`-`REVERT` suggests that the difference in computation happens _before_ the `JUMPI` branching.

Reply:
Another possible issue is perhaps the blocks imported.

``` sh
% grep IMPORTED vyper_flies
DEBUG    eth.chain.chain.Chain:base.py:741 IMPORTED_BLOCK: number 1 | hash 0x38b8b8b3a9ad4aca365148e12c0748b1a8d50b9e8d9d7cba00724311a1036b59
DEBUG    eth.chain.chain.Chain:base.py:741 IMPORTED_BLOCK: number 4 | hash 0xf6e108ea85447da5b99e285cfb9e1466c593559f04ce3512e000d26cee9e55d2
% grep IMPORTED vyper_error
DEBUG    eth.chain.chain.Chain:base.py:556 IMPORTED_BLOCK: number 1 | hash 0x762888e501a157497c8cee2ca541fa6c522830ff75355ee40b41835f54a3cdf8
DEBUG    eth.chain.chain.Chain:base.py:556 IMPORTED_BLOCK: number 1 | hash 0x762888e501a157497c8cee2ca541fa6c522830ff75355ee40b41835f54a3cdf8
```

It would seem that `eth-tester` v0.2 "resets" the block?.. Or tries to build on a previous block?..

Reply:
@veox, thanks for you deep dive. I think your unsubstatiated guess smells right, because this test passes with `eth-tester` v0.1.0b39 (which is pinned to `py-evm` v0.2.0a42), and fails under `eth-tester` v0.2.0b1 (which is pinned to `py-evm` v0.2.0a43), which https://github.com/ethereum/py-evm/pull/1735 looks like the best possible culprit out of the release notes: https://py-evm.readthedocs.io/en/latest/release_notes.html#alpha-43

cc @carver ^

Reply:
Wow @veox, thanks so much for your help! Your debugging skills and depth never ceases to amaze me!

Reply:
Huge thanks to everyone for detection, upstreaming, logging and sleuthing: @fubuloubu @charles-cooper @cburgdorf @veox

That made so it much faster to track this down. I've now got a failing test case in py-evm:

```py
def test_revert_selfdestruct(state):
    state.set_storage(ADDRESS, 1, 2)
    state.persist()

    snapshot = state.snapshot()
    try:
        state.delete_account(ADDRESS)
    finally:
        state.revert(snapshot)

    assert state.get_storage(ADDRESS, 1) == 2
```

I can't figure out why this fairly basic case would not be covered by the ethereum/tests repo (Say, CALL -> SELFDESTRUCT -> REVERT).

I'm still working on tracking down why this is failing, and will post when I've got it solved.

Reply:
https://github.com/ethereum/py-evm/pull/1866

Actually two fixes for similar, related scenarios. (whether the storage was touched before the snapshot or not)